k3s-io/k3s
# Add `kubeadm` style bootstrap token secret support

#### Proposed Changes ####

* Add `k3s token create` - same functionality as `kubeadm token create`
  Can be used to create time-limited join tokens that are automatically cleaned up by the controller-manager when the TTL expires. Tokens are for joining agents only. Servers must know the legacy join/encryption token.
* Allow agents to join using bootstrap tokens or existing kubelet certs (via NodeAuth admission plugin)

#### Types of Changes ####

enhancement

#### Verification ####

* use `k3s token create` to create a join token
  * join an agent to the cluster with the token
* use `k3s token list` to list the token
* use `k3s token delete` to delete the token
* use `k3s token create --ttl 5m` to create a time-limited token
  * note that it is removed when the TTL is up
  * note that nodes joined to the cluster using this token can restart successfully
  * note that nodes cannot rejoin the cluster with the expired token if their node object is deleted from the cluster

#### Testing ####

Covered by CI

#### Linked Issues ####
* https://github.com/k3s-io/k3s/issues/2645
* https://github.com/k3s-io/k3s/issues/6175
* https://github.com/rancher/rke2/issues/3245

#### User-Facing Change ####
```release-note
* K3s now supports `kubeadm` style join tokens. `k3s token create` now creates join token secrets, optionally with a limited TTL.
* K3s agents joined with an expired or deleted token stay in the cluster using existing client certificates via the NodeAuthorization admission plugin, unless their Node object is deleted from the cluster.
```

#### Further Comments ####
